drivers/video/ov5640.c

```diff
@@ -573,12 +659,273 @@ static int ov5640_stream_stop(const struct device *dev)
 }
 
+#define TEST_PATTERN_ENABLE  BIT(7)
+#define TEST_PATTERN_ROLLING BIT(6)
+#define TEST_PATTERN_BAR     (0 << 0)
+#define TEST_PATTERN_SQUARE  (2 << 0)
+
+static const uint8_t test_pattern_val[] = {
+	0,
+	TEST_PATTERN_ENABLE | TEST_PATTERN_BAR | (1 << 2),
+	TEST_PATTERN_ENABLE | TEST_PATTERN_BAR | (1 << 2) | TEST_PATTERN_ROLLING,
+	TEST_PATTERN_ENABLE | TEST_PATTERN_SQUARE,
+	TEST_PATTERN_ENABLE | TEST_PATTERN_SQUARE | TEST_PATTERN_ROLLING,
+};
+
+static int ov5640_set_ctrl_test_pattern(const struct device *dev, int value)
+{
+	const struct ov5640_config *cfg = dev->config;
+
+	if (!IN_RANGE(value, 0, ARRAY_SIZE(test_pattern_val) - 1)) {
+		return -EINVAL;
+	}
+
+	return ov5640_write_reg(&cfg->i2c, PRE_ISP_TEST_SET1, test_pattern_val[value]);
+}
+
+static int ov5640_set_ctrl_hue(const struct device *dev, int value)
+{
+	const struct ov5640_config *cfg = dev->config;
+	int cos_coef, sin_coef, sign = 0;
+
+	if (!IN_RANGE(value, 0, 360)) {
+		return -EINVAL;
+	}
+
+	double rad_val = value;
+	int ret = ov5640_modify_reg(&cfg->i2c, SDE_CTRL0_REG, BIT(0), BIT(0));
+
+	if (ret) {
+		return ret;
+	}
+
+	rad_val = value * PI / 180.0;
+	cos_coef = round(cos(rad_val) * 128);
+	sin_coef = round(sin(rad_val) * 128);
+
+	if (0 <= value && value < 90) {
+		sign = 0x01;
+	} else if (90 <= value && value < 180) {
+		sign = 0x31;
+	} else if (180 <= value && value < 270) {
+		sign = 0x32;
+	} else if (270 <= value && value < 360) {
+		sign = 0x02;
+	}
+
+	struct ov5640_reg hue_params[] = {{SDE_CTRL8_REG, sign},
+					  {SDE_CTRL1_REG, abs(cos_coef)},
+					  {SDE_CTRL2_REG, abs(sin_coef)}};
+
+	return ov5640_write_multi_regs(&cfg->i2c, hue_params, ARRAY_SIZE(hue_params));
+}
+
+static int ov5640_set_ctrl_saturation(const struct device *dev, int value)
+{
+	const struct ov5640_config *cfg = dev->config;
+
+	if (!IN_RANGE(value, 0, UINT8_MAX)) {
+		return -EINVAL;
+	}
+
+	struct ov5640_reg saturation_params[] = {{SDE_CTRL3_REG, value}, {SDE_CTRL4_REG, value}};
+	int ret = ov5640_modify_reg(&cfg->i2c, SDE_CTRL8_REG, BIT(6) | BIT(0), BIT(6) | BIT(0));
+
+	if (ret) {
+		return ret;
+	}
+
+	return ov5640_write_multi_regs(&cfg->i2c, saturation_params, ARRAY_SIZE(saturation_params));
+}
+
+static int ov5640_set_ctrl_brightness(const struct device *dev, int value)
+{
+	const struct ov5640_config *cfg = dev->config;
+
+	if (!IN_RANGE(value, -UINT8_MAX, UINT8_MAX)) {
+		return -EINVAL;
+	}
+
+	struct ov5640_reg brightness_params[] = {{SDE_CTRL8_REG, value >= 0 ? 0x01 : 0x09},
+						 {SDE_CTRL7_REG, abs(value) & 0xff}};
+	int ret = ov5640_modify_reg(&cfg->i2c, SDE_CTRL0_REG, BIT(2), BIT(2));
+
+	if (ret) {
+		return ret;
+	}
+
+	return ov5640_write_multi_regs(&cfg->i2c, brightness_params, ARRAY_SIZE(brightness_params));
+}
+
+static int ov5640_set_ctrl_contrast(const struct device *dev, int value)
+{
+	const struct ov5640_config *cfg = dev->config;
+
+	if (!IN_RANGE(value, 0, UINT8_MAX)) {
+		return -EINVAL;
+	}
```

Comment:
Should it return `-ERANGE` or is `-ERANGE` only for integer overflow?

The real fix would be having some API to query the range rather than trying to guess from the error code anyway.

Reply:
The definition of error values in zephyr (errno.h) is:
- EINVAL: Invalid argument.
- ERANGE: Result too large.

As we check the function argument, therefore EINVAL is more generic and adapted in this case.
Thanks.

Reply:
This is good insight for me on future PRs.